src/maxwell/maxwell_eps.c

```diff
@@ -585,6 +585,12 @@ void set_maxwell_dielectric(maxwell_data *md,
 	     r[1] = i2 * s2;
 	     r[2] = i3 * s3;
 
+       if (mesh_prod == 1) {
+         epsilon(&eps_mean, &eps_inv_mean, r, epsilon_data);
+         md->eps_inv[xyz_index] = eps_inv_mean;
+         goto got_eps_inv;
+       }
```

Comment:
Looks like there is something in this code block which is the cause of NanoComp/meep#2189. Removing this block entirely causes the reported segmentation fault to go away and all the MPB-dependent unit tests in Meep to pass.

Reply:
I think I have an explanation for why this is happening and a fix:

When calling `*fields::get_eigenmode` in `mpb.cpp`, the mpb `mesh_size` is getting set to 1 and `NULL` is passed for the 'epsilon' argument to `set_maxwell_dielectric` in mpb:
```
mpb.cpp 359:  int n[3], local_N, N_start, alloc_N, mesh_size[3] = {1, 1, 1};

mpb.cpp 454: set_maxwell_dielectric(mdata, mesh_size, R, G, NULL, meep_mpb_eps, &eps_data);
```
Thus, when calling mpb _within_ meep it segfaults because epsilon is `NULL`. Checking for epsilon as well as `mesh_prod == 1` solves this and all of the meep tests pass on my machine.

```
CHECK(epsilon || mepsilon, "Neither epsilon nor mepsilon are defined!");
if (epsilon && (mesh_prod == 1)) {
   epsilon(&eps_mean, &eps_inv_mean, r, epsilon_data);
   md->eps_inv[xyz_index] = eps_inv_mean;
   goto got_eps_inv;
}
```

Reply:
Nice catch. However, wouldn't your proposed fix mean that it is *not* possible to disable subpixel smoothing in MPB (the intent of this PR) using its Python API which is (currently) only available via Meep?

(It would be useful to add a unit test for this feature in Meep alongside the other tests in [python/tests/test_mpb.py](https://github.com/NanoComp/meep/blob/master/python/tests/test_mpb.py).)

Reply:
As far as I can tell, when using only the `ModeSolver` class, `epsilon` is always passed to mpb, but when using mpb through meep functions (`EigenmodeSource`, `get_eigenmode`, etc.), only `mepsilon` is passed and `epsilion` is `NULL`. So when using the mpb interface, `epsilon` shouldn't be `NULL` and when using the meep interface, it will be `NULL`. Thus checking for `epsilon` is like checking for whether it's being called within a meep function. 

This code does work on my end and behaves the same as what @flaport has above, with `mesh_size=1` disabling smoothing. 

I think there's more to this though, as there's a specific `meep_mpb_eps` function that's being used as `mepsilon` with the intent (I think) of disabling smoothing as noted in:
```
mpb.cpp 107: return 1; // tells MPB not to do its own subpixel averaging
```
But this might take someone more familiar with the code than I am to dig into deeper.